doc/dev/cephadm/scalability-notes.rst

```diff
@@ -0,0 +1,95 @@
+#############################################
+ Notes and Thoughts on Cephadm's scalability
+#############################################
+
+*********************
+ About this document
+*********************
+
+This document does NOT define a specific proposal or some future work.
+Instead it merely lists a few thoughts that MIGHT be relevant for future
+cephadm enhacements.
+
+*******
+ Intro
+*******
+
+Current situation:
+
+Cephadm manages all registered hosts. This means that it periodically
+scrapes data from each host to identify changes on the host like:
+
+-  disk added/removed
+-  daemon added/removed
+-  host network/firewall etc has changed
+
+Currently, cephadm scrapes each host (up to 10 in parallel) every 6
+minutes, unless a refresh is forced manually.
+
+Refreshes for disks (ceph-volume), daemons (podman/docker), etc, happen
+in sequence.
+
+With the cephadm exporter, we have now reduced the time to scan hosts
+considerably, but the question remains:
+
+Is the cephadm-exporter sufficient to solve all future scalability
+issues?
+
+***********************************************
+ Considerations of cephadm-exporter's REST API
+***********************************************
+
+The cephadm-exporter uses HTTP to serve an endpoint to the hosts
+metadata. We MIGHT encounter some issues with this approach, which need
+to be mitigated at some point.
+
+-  With the cephadm-exporter we use SSH and HTTP to connect to each
+   host. Having two distinct transport layers feels odd, and we might
+   want to consider reducing it to only a single protocol.
+
+-  The current approach of delivering ``bin/cephadm`` to the host doesn't
+   allow the use of external dependencies. This means that we're stuck
+   with the built-in HTTP server lib, which isn't great for providing a
+   good developer experience. ``bin/cephadm`` needs to be packaged and
+   distributed (one way or the other) for us to make use of a better
+   http server library.
+
+************************
+ MON's config-key store
+************************
+
+After the ``mgr/cephadm`` queried metadata from each host, cephadm stores
+the data within the mon's k-v store.
+
+If each host would be allowed to write their own metadata to the store,
+``mgr/cephadm`` would no longer be required to gather the data.
+
+Some questions arise:
+
+-  ``mgr/cephadm`` now needs to query data from the config-key store,
+   instead of relying on cached data.
+
+-  cephadm knows three different types of data: (1) Data that is
+   critical and needs to be stored in the config-key store. (2) Data
+   that can be keept in memeory only. (3) Data that can be stored in
+   Ceph Pool. How can we apply this idea to those different types of
+   data.
```

Comment:
For storing  hosts data i think that we should use mon's k-v store only for a small set of host attributes. basically, the ones that are going allow us to contact with the host and to know what was the last time we had the contact.
Rest of information should be stored in memory. This means to have a quick way to retrieve the data. With the cephadm exporter we can reach probably several hundred of hosts in less than 1 second, so in my view the only thing we need to do is to improve cache refresh and cephadm exporter in order to have different intervals to refresh each type of information... for example:
daemons running in the host each second, disks each 10 minutes.

Reply:
frequent updates to the mon store would likely not be a good idea. There is a potential for the updates from these client(s) to occur at a rater that is faster than the mons are able to persist/coordinate (via PAXOS) ... 

@jecluis correct?

Reply:
to some degree, we have to have the data outside of RADOS, as we're deploying the OSDs  in the first place. And cephadm needs to continue to work without any OSDs running. which means, all data that is required to decide any daemon deployments needs to be part of something independent of RADOS